src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1772,65 +1791,22 @@ InstanceKlass* SystemDictionary::find_or_define_instance_class(Symbol* class_nam
 
   define_instance_class(k, THREAD);
 
-  Handle linkage_exception = Handle(); // null handle
-
   // definer must notify any waiting threads
   {
     MutexLocker mu(THREAD, SystemDictionary_lock);
     PlaceholderEntry* probe = placeholders()->get_entry(name_hash, name_h, loader_data);
     assert(probe != NULL, "DEFINE_CLASS placeholder lost?");
-    if (probe != NULL) {
-      if (HAS_PENDING_EXCEPTION) {
-        linkage_exception = Handle(THREAD,PENDING_EXCEPTION);
-        CLEAR_PENDING_EXCEPTION;
-      } else {
-        probe->set_instance_klass(k);
-      }
-      probe->set_definer(NULL);
-      placeholders()->find_and_remove(name_hash, name_h, loader_data, PlaceholderTable::DEFINE_CLASS, THREAD);
-      SystemDictionary_lock->notify_all();
+    if (!HAS_PENDING_EXCEPTION) {
+      probe->set_instance_klass(k);
     }
-  }
-
-  // Can't throw exception while holding lock due to rank ordering
-  if (linkage_exception() != NULL) {
-    THROW_OOP_(linkage_exception(), NULL); // throws exception and returns
```

Comment:
This rethrows the exception at the end of the function.

Reply:
// definer must notify any waiting threads
  {
    MutexLocker mu(THREAD, SystemDictionary_lock);
    PlaceholderEntry* probe = placeholders()->get_entry(p_index, p_hash, name_h, loader_data);
    assert(probe != NULL, "DEFINE_CLASS placeholder lost?");
    if (probe != NULL) {
      if (HAS_PENDING_EXCEPTION) {
        linkage_exception = Handle(THREAD,PENDING_EXCEPTION);
        CLEAR_PENDING_EXCEPTION;
      } else {
        probe->set_instance_klass(k);
      }
      probe->set_definer(NULL);
      placeholders()->find_and_remove(p_index, p_hash, name_h, loader_data, PlaceholderTable::DEFINE_CLASS, THREAD);
      SystemDictionary_lock->notify_all();
    }
  }

  // Can't throw exception while holding lock due to rank ordering
  if (linkage_exception() != NULL) {
    THROW_OOP_(linkage_exception(), NULL); // throws exception and returns
  }

Reply:
// definer must notify any waiting threads
  {
    MutexLocker mu(THREAD, SystemDictionary_lock);
    PlaceholderEntry* probe = placeholders()->get_entry(name_hash, name_h, loader_data);
    assert(probe != NULL, "DEFINE_CLASS placeholder lost?");
    if (!HAS_PENDING_EXCEPTION) {
      probe->set_instance_klass(k);
    }
    probe->set_definer(NULL);
    placeholders()->find_and_remove(name_hash, name_h, loader_data, PlaceholderTable::DEFINE_CLASS, THREAD);
    SystemDictionary_lock->notify_all();
  }

Reply:
old code vs. new code.

Reply:
Right so my point is that in the new code:

probe->set_definer(NULL);
placeholders()->find_and_remove(p_index, p_hash, name_h, loader_data, PlaceholderTable::DEFINE_CLASS, THREAD);
SystemDictionary_lock->notify_all();

you can now call these while an exception is pending. Whether that is harmless or not depends on the details of that code. I poked around a little and it seems okay but it isn't at all obvious that it is okay.

Reply:
I spent some time reading through this code as well, because I couldn't see why we'd copy the exception.  This is why I changed it, to prevent spending future time reading through this code.  The other reason that I don't like declaring TRAPS for functions that don't throw exceptions is because it's unclear quickly at the call site that the placeholders() call can throw an exception, which would overwrite the one that's already in the _pending_exception field.  It doesn't.  I could have changed the find_and_remove() function to pass THREAD as the first parameter which would make it less suspicious looking, but I had to stop somewhere with this iteration to be nice to code reviewers. Thanks!